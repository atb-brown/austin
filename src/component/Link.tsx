import React from "react";

/**
 * A basic, clickable link for a URL.
 *
 * @param {LinkProps} props Properties.
 * @return {ReactElement} The link component.
 */
export default function Link(props: LinkProps) {
  return (
    <a className="App-link" href={props.url}>
      {props.description}
    </a>
  );
}

type LinkProps = {
  description: string;
  url: string;
};
