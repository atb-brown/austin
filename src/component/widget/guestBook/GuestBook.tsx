import { GuestInfo } from "../../../hook/guests/guestBookInfo";
import useGuestBookInfo from "../../../hook/guests/useGuestBookInfo";
import GuestEntry, { guestEntryGap } from "./GuestEntry";
import React from "react";

/**
 * Create a list of GuestEntry components that will be in the GuestBook.
 *
 * @param {GuestInfo[]} guests The information for each guest.
 * @return {ReactNode[]}
 */
function createGuestEntries(guests: GuestInfo[]): JSX.Element[] {
  const allGuestEntries: JSX.Element[] = [];
  const lastGuestIndex = guests.length - 1;

  for (let guestIndex = 0; guestIndex < guests.length; guestIndex++) {
    const guestInfo = guests[guestIndex];
    allGuestEntries.push(
      <GuestEntry
        isFirst={guestIndex === 0}
        isLast={guestIndex === lastGuestIndex}
        key={guestInfo?.name}
        message={guestInfo?.message}
        name={guestInfo?.name}
        visitDate={guestInfo?.visitDate}
        website={guestInfo?.website}
      />
    );
  }

  return allGuestEntries;
}

/**
 * This component is a "guest book" that displays people that have visited my website and have signed the guest book.
 *
 * @return {ReactElement}
 */
export default function GuestBook(): JSX.Element {
  const { guests } = useGuestBookInfo();

  // TODO: Add a button to "sign my guest book"
  return (
    <div
      data-testid="guest-book"
      style={{
        display: "flex",
        minHeight: "50%",
        width: "99%",
        flexDirection: "column",
        gap: guestEntryGap,
      }}
      title="This component displays everyone who has signed my guest book. It's an exercise in HTTP requests and dependency injection!"
    >
      {createGuestEntries(guests)}
    </div>
  );
}

GuestBook.defaultProps = {
  guestProvider: null,
};
