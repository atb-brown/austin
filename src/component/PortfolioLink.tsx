import Link from "./Link";
import React from "react";

export default function PortfolioLink(props: LinkProps) {
  if (props.issuerDescription && props.issuerUrl) {
    return (
      <LinkWithIssuer
        description={props.description}
        url={props.url}
        urlId={props.urlId}
        issuerDescription={props.issuerDescription}
        issuerUrl={props.issuerUrl}
      />
    );
  }
  return (
    <LinkWithoutIssuer
      description={props.description}
      url={props.url}
      urlId={props.urlId}
    />
  );
}

function LinkWithoutIssuer(props: LinkProps) {
  // TODO: it's a bad idea to have <div/> elements that are children of <p/> elements
  return (
    <div>
      <div>{props.description}</div>
      <div>
        <Link description={props.urlId} url={props.url} />
      </div>
      <br />
    </div>
  );
}

function LinkWithIssuer(props: LinkProps) {
  return (
    <div>
      <div>{props.description}</div>
      <div>
        <Link description={props.urlId} url={props.url} />
      </div>
      <div className="App-soft-text">
        Issued by:{" "}
        <Link
          description={props.issuerDescription ?? ""}
          url={props.issuerUrl ?? ""}
        />
      </div>
      <br />
    </div>
  );
}

type LinkProps = {
  description: string;
  url: string;
  urlId: string;
  issuerDescription?: string;
  issuerUrl?: string;
};
