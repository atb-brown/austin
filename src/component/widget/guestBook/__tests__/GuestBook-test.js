import { wait as guestBook } from "../../../../test/component/GuestBookTestUtils";
import { mockProvider } from "../../../../test/mock/hook/mockUseGuestBookInfo";
import GuestBook from "../GuestBook";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

it("render single guest info to match snapshot", async () => {
  mockProvider();
  const renderResult = render(<GuestBook />);

  await guestBook(renderResult);

  expect(renderResult.asFragment()).toMatchSnapshot();
});

it("render multiple guest info to match snapshot", async () => {
  mockProvider(7);
  const renderResult = render(<GuestBook />);

  await guestBook(renderResult);

  expect(renderResult.asFragment()).toMatchSnapshot();
});

it("render NO guest info to match snapshot", async () => {
  mockProvider(0);
  const renderResult = render(<GuestBook />);

  await guestBook(renderResult);

  expect(renderResult.asFragment()).toMatchSnapshot();
});

afterEach(() => {
  jest.resetAllMocks();
});
