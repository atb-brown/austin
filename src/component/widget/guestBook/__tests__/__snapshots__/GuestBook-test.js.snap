// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render NO guest info to match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="guest-book"
    style="display: flex; min-height: 50%; width: 99%; flex-direction: column; gap: 3px;"
    title="This component displays everyone who has signed my guest book. It's an exercise in HTTP requests and dependency injection!"
  />
</DocumentFragment>
`;

exports[`render multiple guest info to match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="guest-book"
    style="display: flex; min-height: 50%; width: 99%; flex-direction: column; gap: 3px;"
    title="This component displays everyone who has signed my guest book. It's an exercise in HTTP requests and dependency injection!"
  >
    <div
      style="border-top-left-radius: 15px; border-top-right-radius: 15px; border-bottom-left-radius: 0; border-bottom-right-radius: 0; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          austin
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          www.example.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          info loaded
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023
      </div>
    </div>
    <div
      style="border-top-left-radius: 0; border-top-right-radius: 0; border-bottom-left-radius: 0; border-bottom-right-radius: 0; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          person2
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          website2.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          my message 2
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023-01-02
      </div>
    </div>
    <div
      style="border-top-left-radius: 0; border-top-right-radius: 0; border-bottom-left-radius: 0; border-bottom-right-radius: 0; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          person3
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          website3.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          my message 3
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023-01-03
      </div>
    </div>
    <div
      style="border-top-left-radius: 0; border-top-right-radius: 0; border-bottom-left-radius: 0; border-bottom-right-radius: 0; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          person4
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          website4.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          my message 4
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023-01-04
      </div>
    </div>
    <div
      style="border-top-left-radius: 0; border-top-right-radius: 0; border-bottom-left-radius: 0; border-bottom-right-radius: 0; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          person5
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          website5.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          my message 5
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023-01-05
      </div>
    </div>
    <div
      style="border-top-left-radius: 0; border-top-right-radius: 0; border-bottom-left-radius: 0; border-bottom-right-radius: 0; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          person6
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          website6.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          my message 6
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023-01-06
      </div>
    </div>
    <div
      style="border-top-left-radius: 0; border-top-right-radius: 0; border-bottom-left-radius: 15px; border-bottom-right-radius: 15px; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          person7
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          website7.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          my message 7
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023-01-07
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`render single guest info to match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="guest-book"
    style="display: flex; min-height: 50%; width: 99%; flex-direction: column; gap: 3px;"
    title="This component displays everyone who has signed my guest book. It's an exercise in HTTP requests and dependency injection!"
  >
    <div
      style="border-top-left-radius: 15px; border-top-right-radius: 15px; border-bottom-left-radius: 15px; border-bottom-right-radius: 15px; overflow: hidden; display: flex; width: 95%; height: 85px;"
    >
      <div
        class="Guests-Visual"
        style="flex: 2; flex-direction: column; justify-content: space-evenly;"
      >
        <div
          class="Guests-Text-container"
          style="font-size: 75%;"
        >
          austin
        </div>
        <div
          class="Guests-Text-container"
          style="font-size: 50%;"
        >
          www.example.com
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 7; margin-left: 3px; margin-right: 3px;"
      >
        <div
          class="Guests-Text-Container"
          style="font-size: 75%; height: 100%; width: 98%;"
        >
          info loaded
        </div>
      </div>
      <div
        class="Guests-Visual Guests-Text-Container"
        style="flex: 1; font-size: 60%;"
      >
        2023
      </div>
    </div>
  </div>
</DocumentFragment>
`;
